lib/spack/spack/modules/common.py

```diff
@@ -178,7 +178,7 @@ def merge_config_rules(configuration, spec):
         if spec.satisfies(constraint):
             if hasattr(constraint, "override") and constraint.override:
                 spec_configuration = {}
-            update_dictionary_extending_lists(spec_configuration, action)
+            update_dictionary_extending_lists(spec_configuration, copy.deepcopy(action))
```

Comment:
All of those rules are handled in the lines above this change, this just fixes the very technical bug where modulefile generation does not act [as documented](https://spack.readthedocs.io/en/latest/module_file_support.html#selection-by-anonymous-specs) because of mutable objects being mistakenly shared across calls.

Reply:
The actual merge of lists is handled internally by `update_dictionary_extending_lists`, and it alters `spec_configuration` and referenced objects in-place. So doing the `deepcopy` here prevents those merges from being persisted as part of `action` and shared across later modulefile generation calls. Which is the bug presented in #39420.

Reply:
I see, yeah I followed the logic down a couple more layers and I see where the shallow copy was causing the issue now. I am fine with this then.

Reply:
It seems this would be better handled in the called function, since it would be needed at any callsite (it just happens to be the only one for now.